File: android/app/src/main/java/app/provii/wallet/audio/SoundPreset.kt
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust
package app.provii.wallet.audio

/**
 * Sound presets for verification success feedback. Each preset defines musical
 * parameters including note frequencies, ADSR envelope, and harmonic mix ratios
 * for runtime audio synthesis. The [Silent] preset produces no audio and relies
 * on haptic feedback only.
 */
sealed class SoundPreset(
    val name: String,
    val displayName: String,
    val description: String,
    val notes: List<NoteConfig>,
    val envelope: EnvelopeConfig,
    val harmonics: HarmonicConfig,
) {
    /**
     * Configuration for a single synthesized note
     */
    data class NoteConfig(
        val frequency: Float, // Hz
        val startMs: Float, // Start time in milliseconds
        val durationMs: Float, // Duration in milliseconds
        val volume: Float, // Relative volume (0.0 - 1.0)
    )

    /**
     * ADSR envelope configuration
     */
    data class EnvelopeConfig(
        val attackMs: Float, // Time to reach peak
        val sustainLevel: Float, // Sustain level (0.0 - 1.0)
        val releaseRatio: Float, // Release time as ratio of note duration
    )

    /**
     * Harmonic mix ratios for rich tone synthesis
     */
    data class HarmonicConfig(
        val fundamental: Float, // Base tone
        val chorusPlus: Float, // +4 cents detuned
        val chorusMinus: Float, // -4 cents detuned
        val octaveBelow: Float, // f/2
        val octaveAbove: Float, // f*2
        val thirdHarmonic: Float, // f*3
    ) {
        companion object {
            const val CHORUS_DETUNE_CENTS = 4f
        }
    }

    /** Total duration of the sound in milliseconds */
    val totalDurationMs: Float
        get() = notes.maxOfOrNull { it.startMs + it.durationMs } ?: 0f

    /** Whether this preset produces audio (false for silent) */
    val hasAudio: Boolean
        get() = this !is Silent

    // Preset definitions

    object Provii : SoundPreset(
        name = "provii",
        displayName = "Provii",
        description = "The signature Provii chime",
        notes =
            listOf(
                NoteConfig(frequency = 1046.50f, startMs = 0f, durationMs = 60f, volume = 0.65f),
                NoteConfig(frequency = 1318.51f, startMs = 50f, durationMs = 340f, volume = 0.70f),
            ),
        envelope = EnvelopeConfig(attackMs = 4f, sustainLevel = 0.50f, releaseRatio = 0.85f),
        harmonics =
            HarmonicConfig(
                fundamental = 1.0f,
                chorusPlus = 0.15f,
                chorusMinus = 0.15f,
                octaveBelow = 0.25f,
                octaveAbove = 0.0f,
                thirdHarmonic = 0.0f,
            ),
    )

    object Optimal : SoundPreset(
        name = "optimal",
        displayName = "Optimal",
        description = "Balanced, clear confirmation",
        notes =
            listOf(
                NoteConfig(frequency = 523.25f, startMs = 0f, durationMs = 180f, volume = 0.6f),
                NoteConfig(frequency = 783.99f, startMs = 120f, durationMs = 380f, volume = 0.7f),
            ),
        envelope = EnvelopeConfig(attackMs = 8f, sustainLevel = 0.7f, releaseRatio = 0.70f),
        harmonics =
            HarmonicConfig(
                fundamental = 1.0f,
                chorusPlus = 0.40f,
                chorusMinus = 0.40f,
                octaveBelow = 0.50f,
                octaveAbove = 0.15f,
                thirdHarmonic = 0.12f,
            ),
    )

    object Bright : SoundPreset(
        name = "bright",
        displayName = "Bright",
        description = "Higher, more energetic",
        notes =
            listOf(
                NoteConfig(frequency = 659.25f, startMs = 0f, durationMs = 150f, volume = 0.55f),
                NoteConfig(frequency = 987.77f, startMs = 100f, durationMs = 320f, volume = 0.65f),
            ),
        envelope = EnvelopeConfig(attackMs = 8f, sustainLevel = 0.7f, releaseRatio = 0.70f),
        harmonics =
            HarmonicConfig(
                fundamental = 1.0f,
                chorusPlus = 0.40f,
                chorusMinus = 0.40f,
                octaveBelow = 0.35f,
                octaveAbove = 0.15f,
                thirdHarmonic = 0.12f,
            ),
    )

    object Warm : SoundPreset(
        name = "warm",
        displayName = "Warm",
        description = "Lower, more grounded",
        notes =
            listOf(
                NoteConfig(frequency = 392.00f, startMs = 0f, durationMs = 200f, volume = 0.65f),
                NoteConfig(frequency = 587.33f, startMs = 140f, durationMs = 420f, volume = 0.75f),
            ),
        envelope = EnvelopeConfig(attackMs = 12f, sustainLevel = 0.7f, releaseRatio = 0.70f),
        harmonics =
            HarmonicConfig(
                fundamental = 1.0f,
                chorusPlus = 0.40f,
                chorusMinus = 0.40f,
                octaveBelow = 0.60f,
                octaveAbove = 0.15f,
                thirdHarmonic = 0.12f,
            ),
    )

    object Quick : SoundPreset(
        name = "quick",
        displayName = "Quick",
        description = "Short and minimal",
        notes =
            listOf(
                NoteConfig(frequency = 587.33f, startMs = 0f, durationMs = 100f, volume = 0.6f),
                NoteConfig(frequency = 783.99f, startMs = 70f, durationMs = 200f, volume = 0.7f),
            ),
        envelope = EnvelopeConfig(attackMs = 5f, sustainLevel = 0.7f, releaseRatio = 0.70f),
        harmonics =
            HarmonicConfig(
                fundamental = 1.0f,
                chorusPlus = 0.40f,
                chorusMinus = 0.40f,
                octaveBelow = 0.0f,
                octaveAbove = 0.0f,
                thirdHarmonic = 0.0f,
            ),
    )

    object Celebration : SoundPreset(
        name = "celebration",
        displayName = "Celebration",
        description = "Fuller, three-note chime",
        notes =
            listOf(
                NoteConfig(frequency = 523.25f, startMs = 0f, durationMs = 140f, volume = 0.55f),
                NoteConfig(frequency = 659.25f, startMs = 100f, durationMs = 140f, volume = 0.6f),
                NoteConfig(frequency = 783.99f, startMs = 200f, durationMs = 350f, volume = 0.7f),
            ),
        envelope = EnvelopeConfig(attackMs = 8f, sustainLevel = 0.7f, releaseRatio = 0.70f),
        harmonics =
            HarmonicConfig(
                fundamental = 1.0f,
                chorusPlus = 0.40f,
                chorusMinus = 0.40f,
                octaveBelow = 0.50f,
                octaveAbove = 0.15f,
                thirdHarmonic = 0.12f,
            ),
    )

    object Silent : SoundPreset(
        name = "silent",
        displayName = "Silent",
        description = "No sound, haptic only",
        notes = emptyList(),
        envelope = EnvelopeConfig(attackMs = 0f, sustainLevel = 0f, releaseRatio = 0f),
        harmonics =
            HarmonicConfig(
                fundamental = 0f,
                chorusPlus = 0f,
                chorusMinus = 0f,
                octaveBelow = 0f,
                octaveAbove = 0f,
                thirdHarmonic = 0f,
            ),
    )

    companion object {
        /** All available presets */
        val entries: List<SoundPreset> = listOf(Provii, Optimal, Bright, Warm, Quick, Celebration, Silent)

        /** All presets that produce audio (excludes Silent) */
        val audioPresets: List<SoundPreset> = entries.filter { it.hasAudio }

        /** Get preset by name, defaulting to Provii if not found */
        fun fromName(name: String): SoundPreset {
            return entries.find { it.name.equals(name, ignoreCase = true) } ?: Provii
        }
    }
}

File: ios/ProviiWallet/ProviiWallet/Features/Accessibility/Audio/SoundPreset.swift
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust

import Foundation

/// Sound presets for verification success feedback.
/// Each preset defines musical parameters (note frequencies, ADSR envelope, harmonic mix) for synthesised
/// tones played on successful age verification. Includes seven presets from the signature Provii chime to silence.
enum SoundPreset: String, Codable, CaseIterable, Identifiable {
    case provii
    case optimal
    case bright
    case warm
    case quick
    case celebration
    case silent

    var id: String { rawValue }

    // MARK: - Display Properties

    var displayName: String {
        switch self {
        case .provii:      return NSLocalizedString("sound.preset.provii", comment: "Provii")
        case .optimal:     return NSLocalizedString("sound.preset.optimal", comment: "Optimal")
        case .bright:      return NSLocalizedString("sound.preset.bright", comment: "Bright")
        case .warm:        return NSLocalizedString("sound.preset.warm", comment: "Warm")
        case .quick:       return NSLocalizedString("sound.preset.quick", comment: "Quick")
        case .celebration: return NSLocalizedString("sound.preset.celebration", comment: "Celebration")
        case .silent:      return NSLocalizedString("sound.preset.silent", comment: "Silent")
        }
    }

    var description: String {
        switch self {
        case .provii:      return NSLocalizedString("sound.preset.provii.description", comment: "The signature Provii chime")
        case .optimal:     return NSLocalizedString("sound.preset.optimal.description", comment: "Balanced, clear confirmation")
        case .bright:      return NSLocalizedString("sound.preset.bright.description", comment: "Higher, more energetic")
        case .warm:        return NSLocalizedString("sound.preset.warm.description", comment: "Lower, more grounded")
        case .quick:       return NSLocalizedString("sound.preset.quick.description", comment: "Short and minimal")
        case .celebration: return NSLocalizedString("sound.preset.celebration.description", comment: "Fuller, three-note chime")
        case .silent:      return NSLocalizedString("sound.preset.silent.description", comment: "No sound, haptic only")
        }
    }

    // MARK: - Note Configuration

    /// Configuration for a single synthesised note
    struct NoteConfig {
        let frequency: Double      // Hz
        let startMs: Double        // Start time in milliseconds
        let durationMs: Double     // Duration in milliseconds
        let volume: Double         // Relative volume (0.0 - 1.0)
    }

    /// All notes for this preset
    var notes: [NoteConfig] {
        switch self {
        case .provii:
            return [
                NoteConfig(frequency: 1046.50, startMs: 0, durationMs: 60, volume: 0.22),
                NoteConfig(frequency: 1318.51, startMs: 50, durationMs: 340, volume: 0.24)
            ]
        case .optimal:
            return [
                NoteConfig(frequency: 523.25, startMs: 0, durationMs: 180, volume: 0.18),
                NoteConfig(frequency: 783.99, startMs: 120, durationMs: 380, volume: 0.216)
            ]
        case .bright:
            return [
                NoteConfig(frequency: 659.25, startMs: 0, durationMs: 150, volume: 0.16),
                NoteConfig(frequency: 987.77, startMs: 100, durationMs: 320, volume: 0.192)
            ]
        case .warm:
            return [
                NoteConfig(frequency: 392.00, startMs: 0, durationMs: 200, volume: 0.20),
                NoteConfig(frequency: 587.33, startMs: 140, durationMs: 420, volume: 0.24)
            ]
        case .quick:
            return [
                NoteConfig(frequency: 587.33, startMs: 0, durationMs: 100, volume: 0.18),
                NoteConfig(frequency: 783.99, startMs: 70, durationMs: 200, volume: 0.20)
            ]
        case .celebration:
            return [
                NoteConfig(frequency: 523.25, startMs: 0, durationMs: 140, volume: 0.16),
                NoteConfig(frequency: 659.25, startMs: 100, durationMs: 140, volume: 0.176),
                NoteConfig(frequency: 783.99, startMs: 200, durationMs: 350, volume: 0.20)
            ]
        case .silent:
            return []
        }
    }

    /// Total duration of the sound in milliseconds
    var totalDurationMs: Double {
        guard let lastNote = notes.last else { return 0 }
        return lastNote.startMs + lastNote.durationMs
    }

    // MARK: - Envelope Configuration

    /// ADSR envelope configuration
    struct EnvelopeConfig {
        let attackMs: Double       // Time to reach peak
        let sustainLevel: Double   // Sustain level (0.0 - 1.0)
        let releaseRatio: Double   // Release time as ratio of note duration
    }

    var envelope: EnvelopeConfig {
        switch self {
        case .provii:
            return EnvelopeConfig(attackMs: 4, sustainLevel: 0.20, releaseRatio: 0.85)
        case .optimal:
            return EnvelopeConfig(attackMs: 8, sustainLevel: 0.30, releaseRatio: 0.70)
        case .bright:
            return EnvelopeConfig(attackMs: 8, sustainLevel: 0.30, releaseRatio: 0.70)
        case .warm:
            return EnvelopeConfig(attackMs: 12, sustainLevel: 0.30, releaseRatio: 0.70)
        case .quick:
            return EnvelopeConfig(attackMs: 5, sustainLevel: 0.30, releaseRatio: 0.70)
        case .celebration:
            return EnvelopeConfig(attackMs: 8, sustainLevel: 0.30, releaseRatio: 0.70)
        case .silent:
            return EnvelopeConfig(attackMs: 0, sustainLevel: 0, releaseRatio: 0)
        }
    }

    // MARK: - Harmonic Configuration

    /// Harmonic mix ratios for rich tone synthesis
    /// [fundamental, chorus+, chorus-, octaveBelow, octaveAbove, thirdHarmonic]
    struct HarmonicConfig {
        let fundamental: Double      // Base tone
        let chorusPlus: Double       // +4 cents detuned
        let chorusMinus: Double      // -4 cents detuned
        let octaveBelow: Double      // f/2
        let octaveAbove: Double      // f*2
        let thirdHarmonic: Double    // f*3

        static let chorusDetuneCents: Double = 4.0
    }

    var harmonics: HarmonicConfig {
        switch self {
        case .provii:
            return HarmonicConfig(
                fundamental: 1.0,
                chorusPlus: 0.15,
                chorusMinus: 0.15,
                octaveBelow: 0.25,
                octaveAbove: 0.0,
                thirdHarmonic: 0.0
            )
        case .optimal:
            return HarmonicConfig(
                fundamental: 1.0,
                chorusPlus: 0.40,
                chorusMinus: 0.40,
                octaveBelow: 0.50,
                octaveAbove: 0.15,
                thirdHarmonic: 0.12
            )
        case .bright:
            return HarmonicConfig(
                fundamental: 1.0,
                chorusPlus: 0.40,
                chorusMinus: 0.40,
                octaveBelow: 0.35,
                octaveAbove: 0.15,
                thirdHarmonic: 0.12
            )
        case .warm:
            return HarmonicConfig(
                fundamental: 1.0,
                chorusPlus: 0.40,
                chorusMinus: 0.40,
                octaveBelow: 0.60,
                octaveAbove: 0.15,
                thirdHarmonic: 0.12
            )
        case .quick:
            return HarmonicConfig(
                fundamental: 1.0,
                chorusPlus: 0.40,
                chorusMinus: 0.40,
                octaveBelow: 0.0,
                octaveAbove: 0.0,
                thirdHarmonic: 0.0
            )
        case .celebration:
            return HarmonicConfig(
                fundamental: 1.0,
                chorusPlus: 0.40,
                chorusMinus: 0.40,
                octaveBelow: 0.50,
                octaveAbove: 0.15,
                thirdHarmonic: 0.12
            )
        case .silent:
            return HarmonicConfig(
                fundamental: 0,
                chorusPlus: 0,
                chorusMinus: 0,
                octaveBelow: 0,
                octaveAbove: 0,
                thirdHarmonic: 0
            )
        }
    }

    /// Whether this preset produces audio (false for silent)
    var hasAudio: Bool {
        self != .silent
    }
}
